---
title: "Multi-omics Analysis"
output:
  html_document: 
    toc: yes
    fig_width: 12
    fig_height: 8
    code_folding: hide
    warning: FALSE
editor_options: 
  chunk_output_type: console
vignette: >
  %\VignetteIndexEntry{Multi-omics Analysis}
  %\VignetteEngine{knitr::rmarkdown}
---


```{r load_libraries, include=FALSE}

knitr::opts_chunk$set(warning = FALSE)

library(TapestriR)
library(tidyverse)
theme_set(theme_bw())
library(patchwork)


packages <- c("factoextra",  "NbClust", 'uwot', 'ComplexHeatmap')

for(i in 1:length(packages)) {
  package = packages[i]
  if (!requireNamespace(package, quietly = TRUE))
    print(package) #install.packages(package)

  require(package=package, character.only=TRUE)
}

```


# Load multi-omics H5


We will analyze single-cell DNA and protein sequencing data from the Tapestri Platform to demonstrate the ability to detect both CNVs and SNVs simultaneously and see how the expression of proteins changes in different cell populations that are mutant or not. 


```{r load_h5, collapse=FALSE}

#replace filename with path to your data
filename <- system.file("extdata", "4_cell_line_mix_dna_protein.h5", package = "TapestriR")

experiment = read_tap(filename)

experiment
```

# Filter variants

Best practice is to create a multi-assay, multi-sample H5 in Pipeline and apply filters before loading into R. 
We've done some filtering here. Notice how the number of cells and features change after filtering. 

```{r filter_variants}
# Ideally, we would start by loading a filtered H5, but for now, we load data with some basic filters.
filtered_variants = filter_variants(experiment$assays$dna)

# Add VAF layer into the DNA assay.
vaf=round(filtered_variants$data_layers$AD/filtered_variants$data_layers$DP, 3)
vaf[is.na(vaf)] <- 0
filtered_variants = add_data_layer(filtered_variants,'VAF',vaf)

# Add the filtered data back to the experiment. This will be a subset of the rest of the assays to ensure that we have the same cells.
experiment = add_assay(experiment,filtered_variants, keep_common_cells = TRUE)

experiment
```

### Filtered variants

Choosing the right set of variants is critical.

```{r}


# Manually select the rest of variants.
good_variants = experiment$assays$dna$feature_annotations %>% filter(QUAL > 30000)

# Manual filters
manually_filtered_variants = subset_assay(assay = experiment$assays$dna, keep_feature_ids = good_variants$id)
experiment = add_assay(experiment,manually_filtered_variants, keep_common_cells = TRUE)

#DT::datatable(experiment$assays$dna$feature_annotations)
```



# Normalize the read counts

1) Load Protein assay and normalize.
2) Load DNA read counts.

```{r normalize}

# Normalize the protein data using the clr method.
protein_counts_norm = experiment$assays$protein$data_layers$read_counts %>% clr_by_feature() %>% as_tibble(rownames = NA)

# Add the normalized data to the protein assay.
experiment$assays$protein = add_data_layer(experiment$assays$protein,'normalized',protein_counts_norm)

# Normalize the DNA read counts. 
normalized_dna_reads = normalize_dna_reads(experiment$assays$cnv$data_layers$read_counts)
experiment$assays$cnv = add_data_layer(experiment$assays$cnv,'normalized',normalized_dna_reads)

```

# Explore the data {.tabset}

## X-Y plots

1) Select the proteins to plot on the X and Y axes.
2) Select the set of other feature(s) to color the plot by. If you choose more than one feature, each feature will be plotted in a subplot.

```{r fig.width = 12, fig.height = 8}


##################
# Select the proteins to plot on X and Y.
##################

# protein_x = 'CD34'
# protein_y = 'CD38'

protein_x = 1
protein_y = 2

##################
# Select 1 or more features to color by.
# color_by should be a vector of the column header you want to color by.
##################

# All proteins
color_by = experiment$assays$protein$data_layers$normalized

# Select a few proteins.
# color_by =  experiment$assays$protein$data_layers$normalized %>% select('CD110','CD117')

# Select a few variants.
# color_by =  experiment$assays$dna$data_layers$NGT %>% select(1:10) %>% mutate_all(as_factor) %>% mutate_all(recode_genotypes)


p  = tapestri_scatterplot(x = experiment$assays$protein$data_layers$normalized[[protein_x]], 
                 y= experiment$assays$protein$data_layers$normalized[[protein_y]], 
                 color_by = color_by)+ scale_colour_gradient2(low="yellow", mid='grey', high="darkred") 
p = p + xlab(protein_x) + ylab(protein_y) + ggtitle('Color by Protein Expression')
p

# Select a few variants.
color_by =  experiment$assays$dna$data_layers$NGT %>% select(1:10) %>% mutate_all(as_factor) %>% mutate_all(recode_genotypes)


p  = tapestri_scatterplot(x = experiment$assays$protein$data_layers$normalized[[protein_x]], 
                 y= experiment$assays$protein$data_layers$normalized[[protein_y]], 
                 color_by = color_by)
p = p + xlab(protein_x) + ylab(protein_y) + ggtitle('Color by Genotypes')
p

```

## SNV

We will start with the genotype data and also explore how we can identify clones.

### Determine the clones similarly to Tapestri Insights.

Based on genotypes, group each cell into subclones as Tapestri Insights does. 

```{r}

# Define the subclones based on NGT values similar to Tapestri Insights.
ngt_clones = define_subclones(experiment$assays$dna, ignore_zygosity = TRUE)

experiment$assays$dna = add_analysis_layer(assay = experiment$assays$dna, layer_name = 'TI subclone', ngt_clones$subclone_label)

```

### UMAP projection

First, let's reduce the dimensions and see how this will be displayed in a UMAP plot.

```{r projections, message=FALSE, warning=FALSE, cache=TRUE}
data = experiment$assays$dna$data_layers$VAF

projections = list()

# Dimensional reduction using umap.

set.seed(111)
umap_values <- umap(data, scale=TRUE, metric="manhattan", init="laplacian", pca=20) 
projections[['Projection:umap_manhattan DR:pca data:vaf']] = tibble(x = umap_values[,1], y = umap_values[,2])

# umap_values <- umap(data, scale=TRUE, metric="euclidean", init="laplacian", pca=20) 
# projections[['Projection:umap_euclidean DR:pca data:vaf']] = tibble(x = umap_values[,1], y = umap_values[,2])
# 
# umap_values <- umap(data, scale=TRUE, metric="cosine", init="laplacian", pca=20) 
# projections[['Projection:umap_cosine DR:pca data:vaf']] = tibble(x = umap_values[,1], y = umap_values[,2])

experiment$assays$dna = add_analysis_layer(assay = experiment$assays$dna, layer_name = 'projections', as_tibble(projections))

```

### Visualize the projection

```{r umap, message=FALSE, warning=FALSE, cache=TRUE}
# Show a simple plot of the projection.
projection = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`
ggplot(data=projection) + geom_point(aes(x = x, y = y), alpha = 0.5, size=0.8)

```

### Determine the number of clusters

In this step, we determine the optimal number of clusters for our dataset using k-means or hierarchical clustering. This is not a trivial problem. It is left to the user to explore the data. Here are a few examples for determining the number of clusters in your data. 


```{r num_clusters, warning=FALSE, cache=TRUE}

cluster_on = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`

# Elbow method
elbow = fviz_nbclust(cluster_on, kmeans, method = "wss") +
  labs(subtitle = "Elbow method")

# Silhouette method
silhouette = fviz_nbclust(cluster_on, kmeans, method = "silhouette")+
  labs(subtitle = "Silhouette method")

# Gap statistic
# nboot = 50 to keep the function speedy 
# Recommended value: nboot = 500 for your analysis
# Use verbose = FALSE to hide computing progression
# set.seed(123)
# gap_stat = fviz_nbclust(cluster_on, kmeans, nstart = 25,  method = "gap_stat", nboot = 25)+
#   labs(subtitle = "Gap statistic method")


(elbow + silhouette) #/ 
  # (gap_stat + plot_spacer())

```


### Cluster on a UMAP projection

Clustering the data is not a trivial task. 
* Clustering can be done on a UMAP projection or the raw data. 
* There are numerous clustering methods with parameters that need optimization for your data.

It is left to the user to explore the rich community of R packages for clustering. Here, we show clustering by k-means on a UMAP projection.

```{r cluster snv, cache=TRUE}



# Hold all the different customer labels in a single structure.

cluster_by = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`
clusters = list()

# Do the clustering.

for(i in 2:10) {
  kmean_values <- kmeans(cluster_by, i ,iter.max=500)
  clusters[[paste0('umap.kmean.cluster.',i)]] = as_factor(kmean_values$cluster)
}

#############
# Add cluster labels to the analysis data structure.
#############
experiment$assays$dna = add_analysis_layer(assay = experiment$assays$dna, layer_name = 'umap_vaf_clusters', as_tibble(clusters))


```


### Compare the different clusters

In this example, we plot the Tapestri Insights subclone labels on the UMAP projection on VAF data.

color_by can be a data.frame of multiple clustering methods, and all will be plotted side by side.

```{r fig.width = 12, fig.height = 8}

color_by = experiment$assays$dna$analysis_layers$`TI subclone`
projection = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`


p  = tapestri_scatterplot(
                 x = projection$x, 
                 y= projection$y, 
                 color_by = color_by)
p = p + umap_theme() + ggtitle('umap_vaf_clusters')
p

```

### Color the UMAP by genotypes


```{r fig.width = 12, fig.height = 8}
 
 # %>% select(!contains('chr2:198267'))

color_by = experiment$assays$dna$data_layers$NGT %>% select(1:20) %>% mutate_all(as_factor) %>% mutate_all(recode_genotypes)
projection = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`

p  = tapestri_scatterplot(
                 x = projection$x, 
                 y= projection$y, 
                 color_by = color_by)
p = p + umap_theme() 
p = p + ggtitle('Color by Genotypes')
p


```


### Violin plots


```{r fig.width = 12, fig.height = 8}


color_by = experiment$assays$dna$analysis_layers$umap_vaf_clusters$umap.kmean.cluster.4
projection = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`

p  = tapestri_scatterplot(
                 x = projection$x, 
                 y= projection$y, 
                 color_by = color_by)
p = p + umap_theme() + ggtitle('umap_vaf_clusters') + theme(legend.position = 'none')


v = tapestri_violinplot(clusters = color_by,
               features = experiment$assays$protein$data_layers$normalized)
v = v + theme_bw() + theme(legend.position = "none",
                            axis.text.x = element_text(angle = 90, hjust = 1))
  
# pathwork magic
p / v


```


### Color the UMAP by proteins

```{r fig.width = 12, fig.height = 8}

color_by = experiment$assays$protein$data_layers$normalized
projection = experiment$assays$dna$analysis_layers$projections$`Projection:umap_manhattan DR:pca data:vaf`


p  = tapestri_scatterplot(
                 x = projection$x, 
                 y= projection$y, 
                 color_by = color_by)
p = p + umap_theme() + scale_colour_gradient2(low="yellow", mid='grey', high="darkred") 
p = p + ggtitle('Color by Proteins')
p

```

### Single assay heatmaps

Next, we plot the SNVs for each filtered variant across all cells in a heatmap format. We only provide a simple example to get started.

Users should become familiar with ComplexHeatmap::Heatmap https://jokergoo.github.io/ComplexHeatmap-reference/book/. 

```{r}

# Order the features in chr order.
variant_order = experiment$assays$dna$feature_annotations %>%
  mutate(CHROM = as.numeric(CHROM), POS = as.numeric(POS)) %>%
  arrange(CHROM, POS)

genotypes.mat = experiment$assays$dna@data_layers$NGT %>% select(variant_order$id)

clusters = experiment$assays$dna@analysis_layers$umap_vaf_clusters$umap.kmean.cluster.4

snv.h <- ComplexHeatmap::Heatmap(
  as.matrix(genotypes.mat),
  name = "GT",
  col = c("lightgrey", "yellow", "blue", "black"),
  #circlize::colorRamp2(c(0, 1, 2, 3), c("grey", "yellow", "blue", "black"))
  
  heatmap_legend_param = list(labels = c("WT", "HET", "HOM", "Missing")),
  
  split = factor(clusters),
  
  cluster_rows = FALSE,
  show_row_names = FALSE,
  cluster_columns = FALSE,
  row_title_gp = grid::gpar(fontsize = 6),
  column_names_gp = grid::gpar(fontsize = 8),
  show_column_dend = FALSE
)

snv.h

```

## Protein

### Cluster by proteins 
Let's explore our protein count data. 
As we identify 4 clusters using the genotype data, we also define 4 clones. 


```{r cluster_proteins, cache=TRUE}


# Dimensional reduction using umap.
set.seed(111)
umap_values <- umap(experiment$assays$protein$data_layers$normalized, scale=TRUE, metric="euclidean", init="laplacian", pca=20) 

umap_layer = tibble(    
      x = umap_values[,1],
      y = umap_values[,2]
)

experiment$assays$protein = add_analysis_layer(assay = experiment$assays$protein, layer_name = 'umap', umap_layer)
  
# Hold all the different customer labels in a single structure.
# cluster_by = experiment$assays$protein$analysis_layers$umap
cluster_by = experiment$assays$protein$data_layers$normalized

clusters = list()

# Do the clustering.

for(i in 2:10) {
  kmean_values <- kmeans(cluster_by, i ,iter.max=500)
  clusters[[paste0('umap.kmean.cluster.',i)]] = as_factor(kmean_values$cluster)

}

# Add cluster labels to the analysis data structure.
experiment$assays$protein = add_analysis_layer(assay = experiment$assays$protein, layer_name = 'clusters', as_tibble(clusters))


```

### Plot the UMAP and clusters

# Plot all the different clusters on the same UMAP projection.

```{r fig.width = 12, fig.height = 8}


p  = tapestri_scatterplot(
                 x = experiment$assays$protein$analysis_layers$umap$x, 
                 y= experiment$assays$protein$analysis_layers$umap$y, 
                 color_by = experiment$assays$protein$analysis_layers$clusters)

p = p + xlab('') + ylab('')
p = p + umap_theme() 
p

```


### Violin plots

To do:

* Name the clusters based on the signature.
* Show the heatmap of the signature. 

```{r fig.width = 12, fig.height = 8}

p_umap  = tapestri_scatterplot(
                 x = experiment$assays$protein$analysis_layers$umap$x, 
                 y= experiment$assays$protein$analysis_layers$umap$y, 
                 color_by = experiment$assays$protein$analysis_layers$clusters$umap.kmean.cluster.4) + 
  xlab('') + ylab('') + umap_theme() 

p_violin = tapestri_violinplot(
           clusters = experiment$assays$protein$analysis_layers$clusters$umap.kmean.cluster.4,
           features = experiment$assays$protein$data_layers$normalized)

p_umap / p_violin

```


### Color the UMAP by features

```{r fig.width = 12, fig.height = 8}

p  = tapestri_scatterplot(
                 x = experiment$assays$protein$analysis_layers$umap$x, 
                 y= experiment$assays$protein$analysis_layers$umap$y, 
                 color_by = experiment$assays$protein$data_layers$normalized)
p = p + umap_theme() + scale_colour_gradient2(low="yellow", mid='grey', high="darkred") 
p

```
### Single assay heatmaps

To do:

* Provide a way to sort within each cluster.
* Decide what to do with multi-sample.
* Remove features that are not informative.
* Label analytes.

Users should become familiar with ComplexHeatmap::Heatmap https://jokergoo.github.io/ComplexHeatmap-reference/book/. 
We only provide a simple example to get started.

```{r}

# Order the features in chr order.
protein.mat = experiment$assays$protein$data_layers$normalized
clusters = experiment$assays$dna@analysis_layers$umap_vaf_clusters$umap.kmean.cluster.4



protein.h <- ComplexHeatmap::Heatmap(
  as.matrix(protein.mat),
  name = "Protein",
  cluster_rows = FALSE,
  cluster_columns = FALSE,
  
  col = circlize::colorRamp2(c(-2, 0, 2), c("yellow", "grey", "blue")),
  
  split = factor(clusters),
  show_row_names = FALSE,
  row_title_gp = grid::gpar(fontsize = 6),
  #heatmap_legend_param=legend_params,
  column_names_gp = grid::gpar(fontsize = 8),
  show_column_dend = FALSE
)

protein.h
```


## CNV

### Calculate ploidy

Usage: 

* Select a cluster that you think is the normal population.
** Explore your data to identify a control group, which is assumed to be the normal population and, therefore, diploid. 
* Normalize the read counts based on this normal population.
* Plot as a heatmap to review.

```{r}
# Select a clustering method that best represents your data and identify the normal cluster.
# Normalize all the read counts based on the normal cluster.
cnv.mat = compute_ploidy(
  reads = experiment$assays$cnv$data_layers$normalized,
  clusters = experiment$assays$dna$analysis_layers$umap_vaf_clusters$umap.kmean.cluster.4,
  baseline_cluster = 3
 )
experiment$assays$cnv <- add_analysis_layer(assay = experiment$assays$cnv,layer_name = 'norm_to_baseline',data = cnv.mat)

```

### Heatmap

From the heatmap, we can clearly see the regions where there is loss of heterozygosity and ploidy, 2 is reduced closer to 1.

```{r}

cnv.mat = as.matrix(experiment$assays$cnv$analysis_layers$norm_to_baseline)
cnv.mat[is.na(cnv.mat) | is.infinite(cnv.mat)] <-2

clusters = experiment$assays$dna@analysis_layers$umap_vaf_clusters$umap.kmean.cluster.4

cnv.h <- ComplexHeatmap::Heatmap(
   as.matrix(cnv.mat), 
   name = "CNV",
   col = circlize::colorRamp2(c(0, 2, 4), c("yellow", "grey", "blue")), 
   split=factor(clusters),
   
   cluster_rows = FALSE,  
   show_row_names=FALSE, 
   cluster_columns = FALSE,  
   row_title_gp = grid::gpar(fontsize = 6), 
   column_names_gp = grid::gpar(fontsize=8),
   show_column_dend=FALSE)
cnv.h
```

### Ploidy line plot

```{r}

p = tapestri_ploidy_plot(
  normalized_reads = experiment$assays$cnv$analysis_layers$norm_to_baseline,
  clusters = experiment$assays$dna$analysis_layers$umap_vaf_clusters$umap.kmean.cluster.4
  )
p
```

## Multi-omics heatmap 

Combine SNV, CNV, and Protein data into a single heatmap. A nice feature of the ComplexHeatmap package.

```{r  fig.width = 12, fig.height = 8}


snv.h + cnv.h + protein.h


```




